go/src/koding/klient/machine/machinegroup/clients/clients.go

```diff
@@ -0,0 +1,202 @@
+package clients
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"koding/klient/machine"
+
+	"github.com/koding/logging"
+)
+
+// ClientsOpts are the options used to configure set of dynamic clients.
+type ClientsOpts struct {
+	// Builder is a factory used to build clients.
+	Builder machine.ClientBuilder
+
+	// DynAddrInterval indicates how often dynamic client should pull address
+	// function looking for new addresses.
+	DynAddrInterval time.Duration
+
+	// PingInterval indicates how often dynamic client should ping external
+	// machine.
+	PingInterval time.Duration
+
+	// Log is used for logging. If nil, default logger will be created.
+	Log logging.Logger
+}
+
+// Valid checks if provided options are correct.
+func (opts *ClientsOpts) Valid() error {
+	if opts.Builder == nil {
+		return errors.New("nil client builder")
+	}
+	if opts.DynAddrInterval == 0 {
+		return errors.New("dynamic address check interval is not set")
+	}
+	if opts.PingInterval == 0 {
+		return errors.New("ping interval is not set")
+	}
+
+	return nil
+}
+
+// Clients is a set of dynamic clients binded to unique machine ID.
+type Clients struct {
+	builder         machine.ClientBuilder
+	dynAddrInterval time.Duration
+	pingInterval    time.Duration
+
+	log logging.Logger
+
+	mu sync.RWMutex
+	m  map[machine.ID]*machine.DynamicClient
+}
+
+// New creates a new Clients object.
+func New(opts *ClientsOpts) (*Clients, error) {
+	if err := opts.Valid(); err != nil {
+		return nil, err
+	}
+
+	c := &Clients{
+		builder:         opts.Builder,
+		dynAddrInterval: opts.DynAddrInterval,
+		pingInterval:    opts.PingInterval,
+		log:             opts.Log,
+		m:               make(map[machine.ID]*machine.DynamicClient),
+	}
+
+	if c.log == nil {
+		c.log = logging.NewLogger("")
```

Comment:
What about moving this to top-level `var defaultLog = logging.NewLogger("clients")`?